File: typescript/src/crypto.ts
import { createRequire } from "module";

export type WalletFile = Record<string, unknown>;

export type DilithiaAccount = {
  address: string;
  publicKey: string;
  secretKey: string;
  accountIndex: number;
  walletFile?: WalletFile | null;
};

export type DilithiaSignature = {
  algorithm: string;
  signature: string;
};

export type DilithiaKeypair = {
  secretKey: string;
  publicKey: string;
  address: string;
};

export interface DilithiaCryptoAdapter {
  generateMnemonic(): Promise<string>;
  validateMnemonic(mnemonic: string): Promise<void>;
  recoverHdWallet(mnemonic: string): Promise<DilithiaAccount>;
  recoverHdWalletAccount(mnemonic: string, accountIndex: number): Promise<DilithiaAccount>;
  createHdWalletFileFromMnemonic(mnemonic: string, password: string): Promise<DilithiaAccount>;
  createHdWalletAccountFromMnemonic(
    mnemonic: string,
    password: string,
    accountIndex: number
  ): Promise<DilithiaAccount>;
  recoverWalletFile(walletFile: WalletFile, mnemonic: string, password: string): Promise<DilithiaAccount>;
  addressFromPublicKey(publicKeyHex: string): Promise<string>;
  validateAddress(addr: string): Promise<string>;
  addressFromPkChecksummed(publicKeyHex: string): Promise<string>;
  addressWithChecksum(rawAddr: string): Promise<string>;
  validatePublicKey(publicKeyHex: string): Promise<void>;
  validateSecretKey(secretKeyHex: string): Promise<void>;
  validateSignature(signatureHex: string): Promise<void>;
  signMessage(secretKeyHex: string, message: string): Promise<DilithiaSignature>;
  verifyMessage(publicKeyHex: string, message: string, signatureHex: string): Promise<boolean>;
  keygen(): Promise<DilithiaKeypair>;
  keygenFromSeed(seedHex: string): Promise<DilithiaKeypair>;
  seedFromMnemonic(mnemonic: string): Promise<string>;
  deriveChildSeed(parentSeedHex: string, index: number): Promise<string>;
  constantTimeEq(aHex: string, bHex: string): Promise<boolean>;
  hashHex(dataHex: string): Promise<string>;
  setHashAlg(alg: string): Promise<void>;
  currentHashAlg(): Promise<string>;
  hashLenHex(): Promise<number>;
}

export interface SyncDilithiaCryptoAdapter {
  generateMnemonic(): string;
  validateMnemonic(mnemonic: string): void;
  recoverHdWallet(mnemonic: string): DilithiaAccount;
  recoverHdWalletAccount(mnemonic: string, accountIndex: number): DilithiaAccount;
  createHdWalletFileFromMnemonic(mnemonic: string, password: string): DilithiaAccount;
  createHdWalletAccountFromMnemonic(
    mnemonic: string,
    password: string,
    accountIndex: number
  ): DilithiaAccount;
  recoverWalletFile(walletFile: WalletFile, mnemonic: string, password: string): DilithiaAccount;
  addressFromPublicKey(publicKeyHex: string): string;
  validateAddress(addr: string): string;
  addressFromPkChecksummed(publicKeyHex: string): string;
  addressWithChecksum(rawAddr: string): string;
  validatePublicKey(publicKeyHex: string): void;
  validateSecretKey(secretKeyHex: string): void;
  validateSignature(signatureHex: string): void;
  signMessage(secretKeyHex: string, message: string): DilithiaSignature;
  verifyMessage(publicKeyHex: string, message: string, signatureHex: string): boolean;
  keygen(): DilithiaKeypair;
  keygenFromSeed(seedHex: string): DilithiaKeypair;
  seedFromMnemonic(mnemonic: string): string;
  deriveChildSeed(parentSeedHex: string, index: number): string;
  constantTimeEq(aHex: string, bHex: string): boolean;
  hashHex(dataHex: string): string;
  setHashAlg(alg: string): void;
  currentHashAlg(): string;
  hashLenHex(): number;
}

type NativeModuleShape = {
  generateMnemonic: () => string;
  validateMnemonic: (mnemonic: string) => void;
  recoverHdWallet?: (mnemonic: string) => DilithiaAccount;
  recoverHdWalletAccount?: (mnemonic: string, accountIndex: number) => DilithiaAccount;
  createWalletFile?: (password: string) => { mnemonic: string; walletFile: WalletFile };
  createHdWalletFileFromMnemonic?: (mnemonic: string, password: string) => DilithiaAccount;
  createHdWalletAccountFromMnemonic?: (
    mnemonic: string,
    password: string,
    accountIndex: number
  ) => DilithiaAccount;
  recoverWalletFile?: (
    version: number,
    address: string,
    publicKey: string,
    encryptedSk: string,
    nonce: string,
    tag: string,
    accountIndex: number | null,
    mnemonic: string,
    password: string
  ) => DilithiaAccount;
  addressFromPublicKey?: (publicKeyHex: string) => string;
  validateAddress?: (addr: string) => string;
  addressFromPkChecksummed?: (publicKeyHex: string) => string;
  addressWithChecksum?: (rawAddr: string) => string;
  validatePublicKey?: (publicKeyHex: string) => void;
  validateSecretKey?: (secretKeyHex: string) => void;
  validateSignature?: (signatureHex: string) => void;
  signMessage?: (secretKeyHex: string, message: string) => DilithiaSignature;
  verifyMessage?: (publicKeyHex: string, message: string, signatureHex: string) => boolean;
  keygen?: () => { secretKey: string; publicKey: string; address: string };
  keygenFromSeed?: (seedHex: string) => { secretKey: string; publicKey: string; address: string };
  seedFromMnemonic?: (mnemonic: string) => string;
  deriveChildSeed?: (parentSeedHex: string, index: number) => string;
  constantTimeEq?: (aHex: string, bHex: string) => boolean;
  hashHex?: (dataHex: string) => string;
  setHashAlg?: (alg: string) => void;
  currentHashAlg?: () => string;
  hashLenHex?: () => number;
};

function normalizeNativeAccount(account: DilithiaAccount | Record<string, unknown>): DilithiaAccount {
  const source = account as Record<string, unknown>;
  return {
    address: String(source.address ?? ""),
    publicKey: String(source.publicKey ?? source.public_key ?? ""),
    secretKey: String(source.secretKey ?? source.secret_key ?? ""),
    accountIndex: Number(source.accountIndex ?? source.account_index ?? 0),
    walletFile: (source.walletFile ?? source.wallet_file ?? null) as WalletFile | null,
  };
}

export async function loadNativeCryptoAdapter(
  importer: () => Promise<unknown> = () => import("@dilithia/sdk-native")
): Promise<DilithiaCryptoAdapter | null> {
  try {
    const module = (await importer()) as NativeModuleShape;
    return {
      async generateMnemonic() {
        return module.generateMnemonic();
      },
      async validateMnemonic(mnemonic: string) {
        module.validateMnemonic(mnemonic);
      },
      async recoverHdWallet(mnemonic: string) {
        if (!module.recoverHdWallet) {
          throw new Error("Native crypto bridge does not expose recoverHdWallet.");
        }
        return normalizeNativeAccount(module.recoverHdWallet(mnemonic));
      },
      async recoverHdWalletAccount(mnemonic: string, accountIndex: number) {
        if (!module.recoverHdWalletAccount) {
          throw new Error("Native crypto bridge does not expose recoverHdWalletAccount.");
        }
        return normalizeNativeAccount(module.recoverHdWalletAccount(mnemonic, accountIndex));
      },
      async createHdWalletFileFromMnemonic(mnemonic: string, password: string) {
        if (!module.createHdWalletFileFromMnemonic) {
          throw new Error("Native crypto bridge does not expose createHdWalletFileFromMnemonic.");
        }
        return normalizeNativeAccount(module.createHdWalletFileFromMnemonic(mnemonic, password));
      },
      async createHdWalletAccountFromMnemonic(mnemonic: string, password: string, accountIndex: number) {
        if (!module.createHdWalletAccountFromMnemonic) {
          throw new Error("Native crypto bridge does not expose createHdWalletAccountFromMnemonic.");
        }
        return normalizeNativeAccount(
          module.createHdWalletAccountFromMnemonic(mnemonic, password, accountIndex)
        );
      },
      async recoverWalletFile(walletFile: WalletFile, mnemonic: string, password: string) {
        if (!module.recoverWalletFile) {
          throw new Error("Native crypto bridge does not expose recoverWalletFile.");
        }
        return normalizeNativeAccount(
          module.recoverWalletFile(
            Number(walletFile.version ?? 1),
            String(walletFile.address ?? ""),
            String(walletFile.public_key ?? walletFile.publicKey ?? ""),
            String(walletFile.encrypted_sk ?? walletFile.encryptedSk ?? ""),
            String(walletFile.nonce ?? ""),
            String(walletFile.tag ?? ""),
            walletFile.account_index !== undefined && walletFile.account_index !== null
              ? Number(walletFile.account_index)
              : null,
            mnemonic,
            password
          )
        );
      },
      async addressFromPublicKey(publicKeyHex: string) {
        if (!module.addressFromPublicKey) {
          throw new Error("Native crypto bridge does not expose addressFromPublicKey.");
        }
        return module.addressFromPublicKey(publicKeyHex);
      },
      async validateAddress(addr: string) {
        if (!module.validateAddress) {
          throw new Error("Native crypto bridge does not expose validateAddress.");
        }
        return module.validateAddress(addr);
      },
      async addressFromPkChecksummed(publicKeyHex: string) {
        if (!module.addressFromPkChecksummed) {
          throw new Error("Native crypto bridge does not expose addressFromPkChecksummed.");
        }
        return module.addressFromPkChecksummed(publicKeyHex);
      },
      async addressWithChecksum(rawAddr: string) {
        if (!module.addressWithChecksum) {
          throw new Error("Native crypto bridge does not expose addressWithChecksum.");
        }
        return module.addressWithChecksum(rawAddr);
      },
      async validatePublicKey(publicKeyHex: string) {
        if (!module.validatePublicKey) {
          throw new Error("Native crypto bridge does not expose validatePublicKey.");
        }
        module.validatePublicKey(publicKeyHex);
      },
      async validateSecretKey(secretKeyHex: string) {
        if (!module.validateSecretKey) {
          throw new Error("Native crypto bridge does not expose validateSecretKey.");
        }
        module.validateSecretKey(secretKeyHex);
      },
      async validateSignature(signatureHex: string) {
        if (!module.validateSignature) {
          throw new Error("Native crypto bridge does not expose validateSignature.");
        }
        module.validateSignature(signatureHex);
      },
      async signMessage(secretKeyHex: string, message: string) {
        if (!module.signMessage) {
          throw new Error("Native crypto bridge does not expose signMessage.");
        }
        return module.signMessage(secretKeyHex, message);
      },
      async verifyMessage(publicKeyHex: string, message: string, signatureHex: string) {
        if (!module.verifyMessage) {
          throw new Error("Native crypto bridge does not expose verifyMessage.");
        }
        return module.verifyMessage(publicKeyHex, message, signatureHex);
      },
      async keygen() {
        if (!module.keygen) {
          throw new Error("Native crypto bridge does not expose keygen.");
        }
        const raw = module.keygen();
        return { secretKey: raw.secretKey, publicKey: raw.publicKey, address: raw.address };
      },
      async keygenFromSeed(seedHex: string) {
        if (!module.keygenFromSeed) {
          throw new Error("Native crypto bridge does not expose keygenFromSeed.");
        }
        const raw = module.keygenFromSeed(seedHex);
        return { secretKey: raw.secretKey, publicKey: raw.publicKey, address: raw.address };
      },
      async seedFromMnemonic(mnemonic: string) {
        if (!module.seedFromMnemonic) {
          throw new Error("Native crypto bridge does not expose seedFromMnemonic.");
        }
        return module.seedFromMnemonic(mnemonic);
      },
      async deriveChildSeed(parentSeedHex: string, index: number) {
        if (!module.deriveChildSeed) {
          throw new Error("Native crypto bridge does not expose deriveChildSeed.");
        }
        return module.deriveChildSeed(parentSeedHex, index);
      },
      async constantTimeEq(aHex: string, bHex: string) {
        if (!module.constantTimeEq) {
          throw new Error("Native crypto bridge does not expose constantTimeEq.");
        }
        return module.constantTimeEq(aHex, bHex);
      },
      async hashHex(dataHex: string) {
        if (!module.hashHex) {
          throw new Error("Native crypto bridge does not expose hashHex.");
        }
        return module.hashHex(dataHex);
      },
      async setHashAlg(alg: string) {
        if (!module.setHashAlg) {
          throw new Error("Native crypto bridge does not expose setHashAlg.");
        }
        module.setHashAlg(alg);
      },
      async currentHashAlg() {
        if (!module.currentHashAlg) {
          throw new Error("Native crypto bridge does not expose currentHashAlg.");
        }
        return module.currentHashAlg();
      },
      async hashLenHex() {
        if (!module.hashLenHex) {
          throw new Error("Native crypto bridge does not expose hashLenHex.");
        }
        return module.hashLenHex();
      },
    };
  } catch {
    return null;
  }
}

export function loadSyncNativeCryptoAdapter(): SyncDilithiaCryptoAdapter | null {
  try {
    const esmRequire = createRequire(import.meta.url);
    const module = esmRequire("@dilithia/sdk-native") as NativeModuleShape;
    return {
      generateMnemonic() {
        return module.generateMnemonic();
      },
      validateMnemonic(mnemonic: string) {
        module.validateMnemonic(mnemonic);
      },
      recoverHdWallet(mnemonic: string) {
        if (!module.recoverHdWallet) {
          throw new Error("Native crypto bridge does not expose recoverHdWallet.");
        }
        return normalizeNativeAccount(module.recoverHdWallet(mnemonic));
      },
      recoverHdWalletAccount(mnemonic: string, accountIndex: number) {
        if (!module.recoverHdWalletAccount) {
          throw new Error("Native crypto bridge does not expose recoverHdWalletAccount.");
        }
        return normalizeNativeAccount(module.recoverHdWalletAccount(mnemonic, accountIndex));
      },
      createHdWalletFileFromMnemonic(mnemonic: string, password: string) {
        if (!module.createHdWalletFileFromMnemonic) {
          throw new Error("Native crypto bridge does not expose createHdWalletFileFromMnemonic.");
        }
        return normalizeNativeAccount(module.createHdWalletFileFromMnemonic(mnemonic, password));
      },
      createHdWalletAccountFromMnemonic(mnemonic: string, password: string, accountIndex: number) {
        if (!module.createHdWalletAccountFromMnemonic) {
          throw new Error("Native crypto bridge does not expose createHdWalletAccountFromMnemonic.");
        }
        return normalizeNativeAccount(
          module.createHdWalletAccountFromMnemonic(mnemonic, password, accountIndex)
        );
      },
      recoverWalletFile(walletFile: WalletFile, mnemonic: string, password: string) {
        if (!module.recoverWalletFile) {
          throw new Error("Native crypto bridge does not expose recoverWalletFile.");
        }
        return normalizeNativeAccount(
          module.recoverWalletFile(
            Number(walletFile.version ?? 1),
            String(walletFile.address ?? ""),
            String(walletFile.public_key ?? walletFile.publicKey ?? ""),
            String(walletFile.encrypted_sk ?? walletFile.encryptedSk ?? ""),
            String(walletFile.nonce ?? ""),
            String(walletFile.tag ?? ""),
            walletFile.account_index !== undefined && walletFile.account_index !== null
              ? Number(walletFile.account_index)
              : null,
            mnemonic,
            password
          )
        );
      },
      addressFromPublicKey(publicKeyHex: string) {
        if (!module.addressFromPublicKey) {
          throw new Error("Native crypto bridge does not expose addressFromPublicKey.");
        }
        return module.addressFromPublicKey(publicKeyHex);
      },
      validateAddress(addr: string) {
        if (!module.validateAddress) {
          throw new Error("Native crypto bridge does not expose validateAddress.");
        }
        return module.validateAddress(addr);
      },
      addressFromPkChecksummed(publicKeyHex: string) {
        if (!module.addressFromPkChecksummed) {
          throw new Error("Native crypto bridge does not expose addressFromPkChecksummed.");
        }
        return module.addressFromPkChecksummed(publicKeyHex);
      },
      addressWithChecksum(rawAddr: string) {
        if (!module.addressWithChecksum) {
          throw new Error("Native crypto bridge does not expose addressWithChecksum.");
        }
        return module.addressWithChecksum(rawAddr);
      },
      validatePublicKey(publicKeyHex: string) {
        if (!module.validatePublicKey) {
          throw new Error("Native crypto bridge does not expose validatePublicKey.");
        }
        module.validatePublicKey(publicKeyHex);
      },
      validateSecretKey(secretKeyHex: string) {
        if (!module.validateSecretKey) {
          throw new Error("Native crypto bridge does not expose validateSecretKey.");
        }
        module.validateSecretKey(secretKeyHex);
      },
      validateSignature(signatureHex: string) {
        if (!module.validateSignature) {
          throw new Error("Native crypto bridge does not expose validateSignature.");
        }
        module.validateSignature(signatureHex);
      },
      signMessage(secretKeyHex: string, message: string) {
        if (!module.signMessage) {
          throw new Error("Native crypto bridge does not expose signMessage.");
        }
        return module.signMessage(secretKeyHex, message);
      },
      verifyMessage(publicKeyHex: string, message: string, signatureHex: string) {
        if (!module.verifyMessage) {
          throw new Error("Native crypto bridge does not expose verifyMessage.");
        }
        return module.verifyMessage(publicKeyHex, message, signatureHex);
      },
      keygen() {
        if (!module.keygen) {
          throw new Error("Native crypto bridge does not expose keygen.");
        }
        const raw = module.keygen();
        return { secretKey: raw.secretKey, publicKey: raw.publicKey, address: raw.address };
      },
      keygenFromSeed(seedHex: string) {
        if (!module.keygenFromSeed) {
          throw new Error("Native crypto bridge does not expose keygenFromSeed.");
        }
        const raw = module.keygenFromSeed(seedHex);
        return { secretKey: raw.secretKey, publicKey: raw.publicKey, address: raw.address };
      },
      seedFromMnemonic(mnemonic: string) {
        if (!module.seedFromMnemonic) {
          throw new Error("Native crypto bridge does not expose seedFromMnemonic.");
        }
        return module.seedFromMnemonic(mnemonic);
      },
      deriveChildSeed(parentSeedHex: string, index: number) {
        if (!module.deriveChildSeed) {
          throw new Error("Native crypto bridge does not expose deriveChildSeed.");
        }
        return module.deriveChildSeed(parentSeedHex, index);
      },
      constantTimeEq(aHex: string, bHex: string) {
        if (!module.constantTimeEq) {
          throw new Error("Native crypto bridge does not expose constantTimeEq.");
        }
        return module.constantTimeEq(aHex, bHex);
      },
      hashHex(dataHex: string) {
        if (!module.hashHex) {
          throw new Error("Native crypto bridge does not expose hashHex.");
        }
        return module.hashHex(dataHex);
      },
      setHashAlg(alg: string) {
        if (!module.setHashAlg) {
          throw new Error("Native crypto bridge does not expose setHashAlg.");
        }
        module.setHashAlg(alg);
      },
      currentHashAlg() {
        if (!module.currentHashAlg) {
          throw new Error("Native crypto bridge does not expose currentHashAlg.");
        }
        return module.currentHashAlg();
      },
      hashLenHex() {
        if (!module.hashLenHex) {
          throw new Error("Native crypto bridge does not expose hashLenHex.");
        }
        return module.hashLenHex();
      },
    };
  } catch {
    return null;
  }
}

File: python/src/dilithia_sdk/crypto.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from importlib import import_module
from typing import Any, Protocol


WalletFile = dict[str, Any]


@dataclass(slots=True)
class DilithiaAccount:
    address: str
    public_key: str
    secret_key: str
    account_index: int
    wallet_file: WalletFile | None = None


@dataclass(slots=True)
class DilithiaSignature:
    algorithm: str
    signature: str


@dataclass(slots=True)
class DilithiaKeypair:
    secret_key: str
    public_key: str
    address: str


class DilithiaCryptoAdapter(Protocol):
    def generate_mnemonic(self) -> str: ...

    def validate_mnemonic(self, mnemonic: str) -> None: ...

    def recover_hd_wallet(self, mnemonic: str) -> DilithiaAccount: ...

    def recover_hd_wallet_account(self, mnemonic: str, account_index: int) -> DilithiaAccount: ...

    def create_hd_wallet_file_from_mnemonic(self, mnemonic: str, password: str) -> DilithiaAccount: ...

    def create_hd_wallet_account_from_mnemonic(
        self, mnemonic: str, password: str, account_index: int
    ) -> DilithiaAccount: ...

    def recover_wallet_file(self, wallet_file: WalletFile, mnemonic: str, password: str) -> DilithiaAccount: ...

    def address_from_public_key(self, public_key_hex: str) -> str: ...

    def validate_address(self, addr: str) -> str: ...

    def address_from_pk_checksummed(self, public_key_hex: str) -> str: ...

    def address_with_checksum(self, raw_addr: str) -> str: ...

    def validate_public_key(self, public_key_hex: str) -> None: ...

    def validate_secret_key(self, secret_key_hex: str) -> None: ...

    def validate_signature(self, signature_hex: str) -> None: ...

    def sign_message(self, secret_key_hex: str, message: str) -> DilithiaSignature: ...

    def verify_message(self, public_key_hex: str, message: str, signature_hex: str) -> bool: ...

    def keygen(self) -> DilithiaKeypair: ...

    def keygen_from_seed(self, seed_hex: str) -> DilithiaKeypair: ...

    def seed_from_mnemonic(self, mnemonic: str) -> str: ...

    def derive_child_seed(self, parent_seed_hex: str, index: int) -> str: ...

    def constant_time_eq(self, a_hex: str, b_hex: str) -> bool: ...

    def hash_hex(self, data_hex: str) -> str: ...

    def set_hash_alg(self, alg: str) -> None: ...

    def current_hash_alg(self) -> str: ...

    def hash_len_hex(self) -> int: ...


class NativeCryptoAdapter:
    def __init__(self, module: Any) -> None:
        self._module = module

    def generate_mnemonic(self) -> str:
        return self._module.generate_mnemonic()

    def validate_mnemonic(self, mnemonic: str) -> None:
        self._module.validate_mnemonic(mnemonic)

    def recover_hd_wallet(self, mnemonic: str) -> DilithiaAccount:
        payload = self._module.recover_hd_wallet(mnemonic)
        return _normalize_account(payload)

    def recover_hd_wallet_account(self, mnemonic: str, account_index: int) -> DilithiaAccount:
        payload = self._module.recover_hd_wallet_account(mnemonic, account_index)
        return _normalize_account(payload)

    def create_hd_wallet_file_from_mnemonic(self, mnemonic: str, password: str) -> DilithiaAccount:
        payload = self._module.create_hd_wallet_file_from_mnemonic(mnemonic, password)
        return _normalize_account(payload)

    def create_hd_wallet_account_from_mnemonic(
        self, mnemonic: str, password: str, account_index: int
    ) -> DilithiaAccount:
        payload = self._module.create_hd_wallet_account_from_mnemonic(mnemonic, password, account_index)
        return _normalize_account(payload)

    def recover_wallet_file(self, wallet_file: WalletFile, mnemonic: str, password: str) -> DilithiaAccount:
        payload = self._module.recover_wallet_file(
            int(wallet_file.get("version", 1)),
            str(wallet_file.get("address", "")),
            str(wallet_file.get("public_key", wallet_file.get("publicKey", ""))),
            str(wallet_file.get("encrypted_sk", wallet_file.get("encryptedSk", ""))),
            str(wallet_file.get("nonce", "")),
            str(wallet_file.get("tag", "")),
            mnemonic,
            password,
            (
                int(wallet_file["account_index"])
                if wallet_file.get("account_index") is not None
                else None
            ),
        )
        return _normalize_account(payload)

    def address_from_public_key(self, public_key_hex: str) -> str:
        return self._module.address_from_public_key(public_key_hex)

    def validate_address(self, addr: str) -> str:
        return self._module.validate_address(addr)

    def address_from_pk_checksummed(self, public_key_hex: str) -> str:
        return self._module.address_from_pk_checksummed(public_key_hex)

    def address_with_checksum(self, raw_addr: str) -> str:
        return self._module.address_with_checksum(raw_addr)

    def validate_public_key(self, public_key_hex: str) -> None:
        self._module.validate_public_key(public_key_hex)

    def validate_secret_key(self, secret_key_hex: str) -> None:
        self._module.validate_secret_key(secret_key_hex)

    def validate_signature(self, signature_hex: str) -> None:
        self._module.validate_signature(signature_hex)

    def sign_message(self, secret_key_hex: str, message: str) -> DilithiaSignature:
        payload = self._module.sign_message(secret_key_hex, message)
        return DilithiaSignature(
            algorithm=str(payload["algorithm"]),
            signature=str(payload["signature"]),
        )

    def verify_message(self, public_key_hex: str, message: str, signature_hex: str) -> bool:
        return bool(self._module.verify_message(public_key_hex, message, signature_hex))

    def keygen(self) -> DilithiaKeypair:
        payload = self._module.keygen()
        return DilithiaKeypair(
            secret_key=str(payload["secret_key"]),
            public_key=str(payload["public_key"]),
            address=str(payload["address"]),
        )

    def keygen_from_seed(self, seed_hex: str) -> DilithiaKeypair:
        payload = self._module.keygen_from_seed(seed_hex)
        return DilithiaKeypair(
            secret_key=str(payload["secret_key"]),
            public_key=str(payload["public_key"]),
            address=str(payload["address"]),
        )

    def seed_from_mnemonic(self, mnemonic: str) -> str:
        return self._module.seed_from_mnemonic(mnemonic)

    def derive_child_seed(self, parent_seed_hex: str, index: int) -> str:
        return self._module.derive_child_seed(parent_seed_hex, index)

    def constant_time_eq(self, a_hex: str, b_hex: str) -> bool:
        return bool(self._module.constant_time_eq(a_hex, b_hex))

    def hash_hex(self, data_hex: str) -> str:
        return self._module.hash_hex(data_hex)

    def set_hash_alg(self, alg: str) -> None:
        self._module.set_hash_alg(alg)

    def current_hash_alg(self) -> str:
        return self._module.current_hash_alg()

    def hash_len_hex(self) -> int:
        return int(self._module.hash_len_hex())


def _normalize_account(payload: Any) -> DilithiaAccount:
    return DilithiaAccount(
        address=str(payload["address"]),
        public_key=str(payload["public_key"]),
        secret_key=str(payload["secret_key"]),
        account_index=int(payload.get("account_index", 0)),
        wallet_file=payload.get("wallet_file"),
    )


def load_native_crypto_adapter() -> DilithiaCryptoAdapter | None:
    try:
        module = import_module("dilithia_sdk_native")
    except ImportError:
        return None
    return NativeCryptoAdapter(module)


# ---------------------------------------------------------------------------
# Async variants
# ---------------------------------------------------------------------------


class AsyncDilithiaCryptoAdapter(Protocol):
    async def generate_mnemonic(self) -> str: ...

    async def validate_mnemonic(self, mnemonic: str) -> None: ...

    async def recover_hd_wallet(self, mnemonic: str) -> DilithiaAccount: ...

    async def recover_hd_wallet_account(self, mnemonic: str, account_index: int) -> DilithiaAccount: ...

    async def create_hd_wallet_file_from_mnemonic(self, mnemonic: str, password: str) -> DilithiaAccount: ...

    async def create_hd_wallet_account_from_mnemonic(
        self, mnemonic: str, password: str, account_index: int
    ) -> DilithiaAccount: ...

    async def recover_wallet_file(self, wallet_file: WalletFile, mnemonic: str, password: str) -> DilithiaAccount: ...

    async def address_from_public_key(self, public_key_hex: str) -> str: ...

    async def validate_address(self, addr: str) -> str: ...

    async def address_from_pk_checksummed(self, public_key_hex: str) -> str: ...

    async def address_with_checksum(self, raw_addr: str) -> str: ...

    async def validate_public_key(self, public_key_hex: str) -> None: ...

    async def validate_secret_key(self, secret_key_hex: str) -> None: ...

    async def validate_signature(self, signature_hex: str) -> None: ...

    async def sign_message(self, secret_key_hex: str, message: str) -> DilithiaSignature: ...

    async def verify_message(self, public_key_hex: str, message: str, signature_hex: str) -> bool: ...

    async def keygen(self) -> DilithiaKeypair: ...

    async def keygen_from_seed(self, seed_hex: str) -> DilithiaKeypair: ...

    async def seed_from_mnemonic(self, mnemonic: str) -> str: ...

    async def derive_child_seed(self, parent_seed_hex: str, index: int) -> str: ...

    async def constant_time_eq(self, a_hex: str, b_hex: str) -> bool: ...

    async def hash_hex(self, data_hex: str) -> str: ...

    async def set_hash_alg(self, alg: str) -> None: ...

    async def current_hash_alg(self) -> str: ...

    async def hash_len_hex(self) -> int: ...


class AsyncNativeCryptoAdapter:
    """Wraps a synchronous :class:`NativeCryptoAdapter` and delegates every
    call to a thread executor via :func:`asyncio.to_thread` so that
    CPU-intensive native crypto work never blocks the event loop.
    """

    def __init__(self, sync_adapter: NativeCryptoAdapter) -> None:
        self._sync = sync_adapter

    async def generate_mnemonic(self) -> str:
        return await asyncio.to_thread(self._sync.generate_mnemonic)

    async def validate_mnemonic(self, mnemonic: str) -> None:
        await asyncio.to_thread(self._sync.validate_mnemonic, mnemonic)

    async def recover_hd_wallet(self, mnemonic: str) -> DilithiaAccount:
        return await asyncio.to_thread(self._sync.recover_hd_wallet, mnemonic)

    async def recover_hd_wallet_account(self, mnemonic: str, account_index: int) -> DilithiaAccount:
        return await asyncio.to_thread(self._sync.recover_hd_wallet_account, mnemonic, account_index)

    async def create_hd_wallet_file_from_mnemonic(self, mnemonic: str, password: str) -> DilithiaAccount:
        return await asyncio.to_thread(self._sync.create_hd_wallet_file_from_mnemonic, mnemonic, password)

    async def create_hd_wallet_account_from_mnemonic(
        self, mnemonic: str, password: str, account_index: int
    ) -> DilithiaAccount:
        return await asyncio.to_thread(
            self._sync.create_hd_wallet_account_from_mnemonic, mnemonic, password, account_index
        )

    async def recover_wallet_file(self, wallet_file: WalletFile, mnemonic: str, password: str) -> DilithiaAccount:
        return await asyncio.to_thread(self._sync.recover_wallet_file, wallet_file, mnemonic, password)

    async def address_from_public_key(self, public_key_hex: str) -> str:
        return await asyncio.to_thread(self._sync.address_from_public_key, public_key_hex)

    async def validate_address(self, addr: str) -> str:
        return await asyncio.to_thread(self._sync.validate_address, addr)

    async def address_from_pk_checksummed(self, public_key_hex: str) -> str:
        return await asyncio.to_thread(self._sync.address_from_pk_checksummed, public_key_hex)

    async def address_with_checksum(self, raw_addr: str) -> str:
        return await asyncio.to_thread(self._sync.address_with_checksum, raw_addr)

    async def validate_public_key(self, public_key_hex: str) -> None:
        await asyncio.to_thread(self._sync.validate_public_key, public_key_hex)

    async def validate_secret_key(self, secret_key_hex: str) -> None:
        await asyncio.to_thread(self._sync.validate_secret_key, secret_key_hex)

    async def validate_signature(self, signature_hex: str) -> None:
        await asyncio.to_thread(self._sync.validate_signature, signature_hex)

    async def sign_message(self, secret_key_hex: str, message: str) -> DilithiaSignature:
        return await asyncio.to_thread(self._sync.sign_message, secret_key_hex, message)

    async def verify_message(self, public_key_hex: str, message: str, signature_hex: str) -> bool:
        return await asyncio.to_thread(self._sync.verify_message, public_key_hex, message, signature_hex)

    async def keygen(self) -> DilithiaKeypair:
        return await asyncio.to_thread(self._sync.keygen)

    async def keygen_from_seed(self, seed_hex: str) -> DilithiaKeypair:
        return await asyncio.to_thread(self._sync.keygen_from_seed, seed_hex)

    async def seed_from_mnemonic(self, mnemonic: str) -> str:
        return await asyncio.to_thread(self._sync.seed_from_mnemonic, mnemonic)

    async def derive_child_seed(self, parent_seed_hex: str, index: int) -> str:
        return await asyncio.to_thread(self._sync.derive_child_seed, parent_seed_hex, index)

    async def constant_time_eq(self, a_hex: str, b_hex: str) -> bool:
        return await asyncio.to_thread(self._sync.constant_time_eq, a_hex, b_hex)

    async def hash_hex(self, data_hex: str) -> str:
        return await asyncio.to_thread(self._sync.hash_hex, data_hex)

    async def set_hash_alg(self, alg: str) -> None:
        await asyncio.to_thread(self._sync.set_hash_alg, alg)

    async def current_hash_alg(self) -> str:
        return await asyncio.to_thread(self._sync.current_hash_alg)

    async def hash_len_hex(self) -> int:
        return await asyncio.to_thread(self._sync.hash_len_hex)


def load_async_native_crypto_adapter() -> AsyncDilithiaCryptoAdapter | None:
    try:
        module = import_module("dilithia_sdk_native")
    except ImportError:
        return None
    return AsyncNativeCryptoAdapter(NativeCryptoAdapter(module))
